Ruins/src/main/java/atomicstryker/ruins/common/RuinTemplateRule.java

```diff
@@ -1461,6 +1457,51 @@ private static TileEntity realizeBlock(World world, int x, int y, int z, Block b
         return entity;
     }
 
+    // parsing block property specifications can be expensive! cache states to avoid repeated calculation
+    @SuppressWarnings("deprecation")
+    private IBlockState getCachedBlockState(Block block, int blocknum)
+    {
+        // use cached state; if no metadata, use default state
+        IBlockState state = blocknum != UNSPECIFIED_BLOCKNUM ? blockBSs[blocknum] : block.getDefaultState();
+
+        // if no cached state, interpret metadata string and save result for next time
+        if (state == null)
+        {
+            // convert data value to state
+            String metadata = blockMDs[blocknum];
+            if (metadata != null && !metadata.isEmpty())
+            {
+                try
+                {
+                    int value = Integer.parseInt(metadata);
+                    if ((value & ~0b1111) == 0)

```

Comment:
document this bit filter for mere mortals

Reply:
Ha! Knew you'd like that.